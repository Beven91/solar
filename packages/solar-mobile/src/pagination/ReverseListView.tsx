/**
* @module ReverseListView
* @description 反向滚动的ListView组件
*/
import React from 'react';
import { Icon } from 'antd-mobile';

export interface ReverseListViewProps {
  // 当前页码值
  page: number
  // 原始行数据
  originalRows: Array<any>
  // class样式
  className?: string
  // 渲染每一条数据
  renderRow: (row: any, id: any, rowId?: any, hightlightRow?: boolean) => React.ReactElement<any>
  // 当滚动到底部时触发次函数，用于执行分页操作
  onEndReached: () => void
  // 当滚动顶部出现的内容
  renderHeader?: () => React.ReactElement<any>
  // 是否有更多数据
  hasMore: boolean,
  // 调用onEndReached之前的临界值，单位是像素
  onEndReachedThreshold?: number
}

export default class ReverseListView extends React.Component<ReverseListViewProps, {}> {
  // 默认属性值
  static defaultProps = {
    originalRows: [] as Array<any>,
    onEndReachedThreshold: 20,
    renderHeader: () => (
      <div className="reverse-header">
        <div className="scroll-end-loading">
          <Icon type="loading" className="loading-icon" size="xxs" />
          <span className="loading-text">加载中，请稍后....</span>
        </div>
      </div>
    ),
  }

  private shouldSync: boolean

  private isPaginating: boolean

  private lastScrollTop: number

  private prevScrollHeight: number

  // 滚动容器ref
  containerRef = React.createRef<HTMLDivElement>()

  // 获取当前滚动容器的scrollHeight
  get scrollHeight() {
    return this.getInnerViewNode().scrollHeight;
  }

  // 获取内部dom节点
  getInnerViewNode() {
    return this.containerRef.current;
  }

  // 滚动指定位置
  scrollTo(x: number, y: number) {
    const current = this.getInnerViewNode();
    current.scrollTop = y;
    current.scrollLeft = x || 0;
  }

  // 滚动事件
  handleScroll = (e: React.UIEvent<HTMLDivElement, UIEvent>) => {
    const { hasMore, onEndReached, onEndReachedThreshold } = this.props;
    if (this.isPaginating || !hasMore) {
      return;
    }
    const element = e.target as HTMLDivElement;
    const scrollTop = element.scrollTop;
    this.lastScrollTop = scrollTop;
    if (scrollTop < onEndReachedThreshold) {
      this.isPaginating = true;
      element.classList.add('no-scroll');
      onEndReached();
    }
  }

  // 处理滚动位置
  handlePosition(prevScrollHeight: number) {
    const { page } = this.props;
    const scrollHeight = this.scrollHeight;
    const y = page === 1 ? scrollHeight : this.lastScrollTop + (scrollHeight - prevScrollHeight);
    this.scrollTo(0, y);
    this.containerRef.current.classList.remove('no-scroll');
  }

  /**
   * 当页码值发生变化时，意味着新一页的数据已经请求完毕，这里开始渲染新一页的行数据
   * @param {*}} nextProps 新的属性s
   */
  componentWillReceiveProps(nextProps: ReverseListViewProps) {
    const { originalRows } = nextProps;
    if (!this.shouldSync && originalRows.length !== this.props.originalRows.length) {
      this.prevScrollHeight = this.scrollHeight;
      this.shouldSync = true;
      this.isPaginating = false;
    }
  }

  componentDidUpdate() {
    if (this.shouldSync) {
      this.handlePosition(this.prevScrollHeight);
      this.shouldSync = false;
    }
  }

  renderRows() {
    const { originalRows, renderRow } = this.props;
    const rows = [];
    for (let i = originalRows.length - 1; i >= 0; i--) {
      const row = originalRows[i];
      rows.push(<div className={`row-${i}`} key={`row-${i}`}>{renderRow(row, i)}</div>);
    }
    return rows;
  }

  // 渲染组件
  render() {
    const { className, hasMore, renderHeader } = this.props;
    return (
      <div
        ref={this.containerRef}
        onScroll={this.handleScroll}
        className={`reverser-scroll ${className}`}
      >
        {hasMore ? renderHeader() : ''}
        <div>
          {this.renderRows()}
        </div>
      </div>
    );
  }
}
