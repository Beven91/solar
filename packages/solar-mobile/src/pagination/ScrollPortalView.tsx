/**
 * 名称：ScrollView 转移视图
 * 日期：2018-03-14
 * 描述：用于实现滚动容器自定义
 *      部分情况下需要的滚动容器，不是分页列表本身，而是所处的父级dom
 *      可以使用此组件来实现一个【滚动转移】ScrollView
 * 本组件大部分代码参照antd-mobile ListView官方引用的模块 ScrollView
 * https://github.com/react-component/m-list-view/blob/master/src/ScrollView.js
 */

import React from 'react';

export interface ScrollPortalViewProps {
  // 其他属性来源于ListView组件
  // 自定义滚动容器 返回实际出现滚动条的dom元素 例如: ()=>document.getElementById('ss');
  getScrollContainer?: () => HTMLElement,
  dataSource: any,
  className?: string
  horizontal?: boolean
  onScroll?: (e: any, metrics: any) => void,
  initialListSize?: number
}

export default class ScrollPortalView extends React.Component<ScrollPortalViewProps, {}> {
  public isRemovedScroll: boolean

  /**
   * 获取当前滚动容器
   */
  get scroller() {
    return this.props.getScrollContainer();
  }

  /**
   * 获取当前滚动容器上下文参数
   */
  getMetrics = () => {
    const isVertical = !this.props.horizontal;
    return {
      visibleLength: 0,
      contentLength: this.scroller[isVertical ? 'scrollHeight' : 'scrollWidth'],
      offset: this.scroller[isVertical ? 'scrollTop' : 'scrollLeft'],
    };
  }

  onScroll = (e: any) => {
    const { onScroll } = this.props;
    onScroll(e, this.getMetrics());
  }

  /**
   * 组件创建完毕,这里执行滚动事件绑定
   */
  componentDidMount() {
    setTimeout(() => {
      if (this.scroller) {
        this.scroller.addEventListener('scroll', this.onScroll);
      }
    }, 50);
  }

  /**
   * 组件更新时
   * 来源:https://github.com/react-component/m-list-view/blob/master/src/ScrollView.js
   * 问题情景：用户滚动内容后，改变 dataSource 触发 ListView componentWillReceiveProps
   * 内容变化后 scrollTop 如果改变、会自动触发 scroll 事件，而此事件应该避免被执行
   */
  componentWillUpdate(nextProps: ScrollPortalViewProps) {
    if ((this.props.dataSource !== nextProps.dataSource ||
      this.props.initialListSize !== nextProps.initialListSize) && this.scroller) {
      this.isRemovedScroll = true;
      this.scroller.removeEventListener('scroll', this.onScroll);
    }
  }

  /**
   * 组件更新完毕
   * 重新尝试绑定滚动事件
   */
  componentDidUpdate() {
    if (this.isRemovedScroll) {
      setTimeout(() => {
        this.isRemovedScroll = false;
        this.scroller.addEventListener('scroll', this.onScroll);
      }, 0);
    }
  }

  /**
   * 组件销毁
   * 取消事件绑定
   */
  componentWillUnmount() {
    if (this.scroller) {
      this.scroller.removeEventListener('scroll', this.onScroll);
    }
  }

  /**
   * 渲染组件
   */
  render() {
    const { className, children } = this.props;
    return <div className={className} >{children}</div>;
  }
}
