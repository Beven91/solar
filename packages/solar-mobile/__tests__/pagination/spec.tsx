import React from 'react';
import Pagination, { EndReachedEvent } from '../../src/pagination';
import ScrollPortalView from '../../src/pagination/ScrollPortalView';
import ReverseListView from '../../src/pagination/ReverseListView';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.useFakeTimers();

class App extends React.Component<any, any> {
  public onEndReached: jest.Mock

  public pagination: Pagination

  constructor(props: any) {
    super(props);
    this.onEndReached = jest.fn();
    this.onEndReached.mockImplementation(this.onEndReached2);
  }

  state = {
    data: [] as any,
    hasMore: true,
  }

  renderRow = jest.fn().mockImplementation((row) => {
    return <div>{row.name}</div>;
  })

  onEndReached2 = (e: EndReachedEvent) => {
    const rows = getPageRows(e.page, e.size);
    const data = this.state.data;
    const increment = this.props.increment;
    const allRows = increment ? data.concat(data) : rows;
    const { onPaginating } = this.props;
    onPaginating && onPaginating();
    this.setState({
      data: [...allRows],
      hasMore: rows.length > 0,
    });
  }

  componentDidUpdate() {
    jest.runOnlyPendingTimers();
  }

  render() {
    return (
      <Pagination
        key="order-list"
        className="order-list"
        {...this.props}
        ref={(pagination) => this.pagination = pagination}
        hasMore={this.state.hasMore}
        dataSource={this.state.data}
        pageSize={3}
        increment={this.props.increment}
        renderRow={this.renderRow}
        onEndReached={this.onEndReached}
        getScrollContainer={this.props.getScrollContainer}
      />
    );
  }
}

describe('Pagination', () => {
  it('pagination', () => {
    const tree = mount(
      <App direction="bottom-to-top" />
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('pagination-revese', () => {
    const tree = mount(
      <App />
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });


  it('pagination increment', () => {
    const tree = mount(
      <App
        filter={(row1: any, row2: any) => row1 !== row2}
        increment={false}
        getScrollContainer={() => document.body}
      />
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();

    // 手动触发分页
    const instance = tree.instance() as App;
    instance.pagination.onEndReached();

    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });


  it('pagination custom scroller', () => {
    const tree = mount(
      <App getScrollContainer={() => document.body} />
    );
    const instance = tree.instance() as App;
    const onEndReached = instance.onEndReached;

    // 断言：在当前状态下，渲染的数据为 3条
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();

    // 场景一：没有滚动到底部，不触发滚动分页
    Object.defineProperty(document.body, 'scrollHeight', { configurable: true, value: 1500 });
    document.body.scrollTop = 100;
    document.body.dispatchEvent(new Event('scroll'));
    // 由于没有滚动到底部，所以这里不会触发滚动分页
    // 值为什么是1? 因为默认初始化会加载一次数据，所以调用了一次
    expect(onEndReached).toHaveBeenCalledTimes(1);
    // 断言：在没有触发滚动分页状态下，渲染数据仍然为3条
    expect(toJson(tree.render(), { noKey: true, mode: 'deep' })).toMatchSnapshot();

    // 场景二：滚动到底部，触发滚动分页
    document.body.scrollTop = 890;
    document.body.dispatchEvent(new Event('scroll'));
    expect(onEndReached).toHaveBeenCalledTimes(2);
    expect(toJson(tree.render(), { noKey: true, mode: 'deep' })).toMatchSnapshot();
    // 触发滚动分页

    // 销毁组件
    tree.unmount();
  });

  it('pagination.refresh', () => {
    const tree = mount(
      <App getScrollContainer={() => document.body} />
    );
    const instance = tree.instance() as App;
    const onEndReached = instance.onEndReached;
    // 滚动分页
    document.body.scrollTop = 890;
    document.body.dispatchEvent(new Event('scroll'));

    // 获取分页组件实例
    const pagination = tree.find<Pagination>(Pagination).instance();
    // 断言：当前页码应该在第二页
    expect(pagination.state.page).toBe(2);

    // 清除调用记录
    onEndReached.mockClear();
    // 调用刷新函数
    // pagination.refresh();

    // 断言：当前页码应该重回到第一页状态
    // expect(pagination.state.page).toBe(1);
    // 断言：onEndReached 需要被调用，进行一次数据刷新
    // expect(onEndReached).toHaveBeenCalledTimes(1);
  });


  it('pagination.componentWillReceiveProps', () => {
    const tree = mount(
      <Pagination
        dataSource={[]}
        onEndReached={() => []}
        renderRow={() => <div></div>}
        increment={false}
      />
    );
    tree.setProps({ data: null });
    expect(toJson(tree.render(), { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('ScrollView', () => {
    const dataSource = [] as any;
    const getScrollContainer = jest.fn();
    getScrollContainer.mockImplementation(() => document.body);
    const scrollView = mount(
      <ScrollPortalView
        dataSource={dataSource}
        horizontal={true}
        getScrollContainer={getScrollContainer}
      />
    );
    const instance = scrollView.instance() as ScrollPortalView;
    instance.getMetrics();
    scrollView.setProps({ initialListSize: 20 });
    scrollView.setProps({ horizontal: false });

    instance.isRemovedScroll = false;
    instance.componentDidUpdate();
    scrollView.update();
    jest.runOnlyPendingTimers();
    getScrollContainer.mockImplementation(() => null);

    scrollView.instance().componentDidMount();
    jest.runOnlyPendingTimers();

    scrollView.unmount();
  });

  /**
   * =================>
   *
   * 逆向滚动 测试用例
   *
   * =================>
   */

  it('pagination bottom-to-top', () => {
    // 逆向滚动分页
    const tree = mount(
      <App
        className="reverse-pagination"
        direction="bottom-to-top"
      />
    );
    const instance = tree.instance() as App;
    const onEndReached = instance.onEndReached;
    const renderRow = instance.renderRow;
    // 断言：初始状态下 逆向滚动分页镜像匹配
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();

    // 断言：onEndReached 会被调用一次，（初始化的分页请求)
    expect(onEndReached).toHaveBeenCalledTimes(1);

    // 清除调用记录
    onEndReached.mockClear();

    // 获取滚动容器
    const scroller = tree.find('.reverse-pagination').first();
    // 模拟滚动变化
    simulateScroll(scroller, { scrollTop: 180 });
    jest.runOnlyPendingTimers();
    // 断言：此时不会触发onEndReached 默认 当scrollTop < 10 时才会触发分页请求
    expect(onEndReached).not.toHaveBeenCalled();

    // 模拟滚动到 20以内，
    simulateScroll(scroller, { scrollTop: 18 });
    jest.runOnlyPendingTimers();
    // 断言：当滚动到顶部，会触发分页请求
    expect(onEndReached).toHaveBeenCalledTimes(1);

    // 断言renderRow 会被调用
    expect(renderRow).toHaveBeenCalled();
    // 清除renderRow 调用记录
    renderRow.mockClear();
    // 在不改变page的情况下，强制触发组件刷新
    tree.instance().forceUpdate();
    // 断言：renderRow 不会被调用
    // 预期：在不改变page的情况下，不会触发ReverseListView 组件prependRows 这样可以防止重复行渲染
    expect(renderRow).not.toHaveBeenCalled();

    // 设置当前无更多记录，hasMore=false
    tree.setState({ hasMore: false });
    // 清除调用记录
    onEndReached.mockClear();
    // 模拟滚动到 10以内，
    simulateScroll(scroller, { scrollTop: 9 });
    jest.runOnlyPendingTimers();
    // 断言：由于hasMore=false 所以后续滚动不会触发分页
    expect(onEndReached).not.toHaveBeenCalled();
    // 断言：在无更多分页数据状态下，不会出现loading头部
    expect(tree.find('.reverse-header').length).toBe(0);
  });

  it('pagination bottom-to-top handlePosition', () => {
    // 逆向滚动分页
    const tree = mount(
      <App
        className="reverse-pagination"
        direction="bottom-to-top"
      />
    );
    // 获取滚动容器
    const scroller = tree.find('.reverse-pagination').first();
    const reverser = tree.find(ReverseListView);
    const reverserInstance = reverser.instance() as ReverseListView;
    const dom = scroller.getDOMNode();
    const originalHandlePosition = reverserInstance.handlePosition;

    // 模拟当前分页的内容总体高度为 1200
    Object.defineProperty(dom, 'scrollHeight', { value: 1200, configurable: true });
    // 断言：dom.scrollHeight 为 1200
    expect(dom.scrollHeight).toBe(1200);

    // 通过覆写handlePosition 来测试scrollTop的值是否准确
    reverserInstance.handlePosition = (...options) => {
      // 模拟在请求数据完成后，scrollHeight 变为 2400
      Object.defineProperty(dom, 'scrollHeight', { value: 2400, configurable: true });
      originalHandlePosition.apply(reverserInstance, [...options]);
    };

    // 模拟滚动分页
    simulateScroll(scroller, { scrollTop: 9 });
    jest.runOnlyPendingTimers();

    // 断言：scrollTop 值应该为 1209
    //  即: scrollHeight(2400) - prev ScrollHeight(1200) + last scrollTop(9)
    //  即: 2400 - 1200 + 9 = 1209
    // 原因：在滚动顶部，分页请求完毕，且渲染完毕后，需要将滚动容器的scrollTop设置到最后滚动的位置
    // 否则，在渲染后，顶部显示的是最后一页的最后几条记录，这样位置不延续
    expect(dom.scrollTop).toBe(1209);
  });

  it('pagination bottom-to-top repeat-scroll', () => {
    const onPaginating = jest.fn();
    // 逆向滚动分页
    const tree = mount(
      <App
        className="reverse-pagination"
        onPaginating={onPaginating}
        direction="bottom-to-top"
      />
    );
    const instance = tree.instance() as App;
    const onEndReached = instance.onEndReached;
    // 获取滚动容器
    const scroller = tree.find('.reverse-pagination').first();
    // 清除调用记录
    onEndReached.mockClear();

    // 模拟滚动分页
    simulateScroll(scroller, { scrollTop: 9 });
    // 此时在 onPaginating时，模拟再一次滚动
    onPaginating.mockImplementation(() => {
      // 模拟滚动分页
      simulateScroll(scroller, { scrollTop: 8 });
    });
    jest.runOnlyPendingTimers();

    // 断言：虽然连续滚动了两次，但是onEndReached 仅需要被调用一次
    expect(onEndReached).toHaveBeenCalledTimes(1);
  });
});

function simulateScroll(wrapper: any, config: any) {
  const node = wrapper.getDOMNode();
  node.scrollTop = config.scrollTop;
  wrapper.simulate('scroll', { target: node });
}

function getPageRows(page: number, size: number) {
  const rows = [];
  if (page > 2) {
    return [];
  }
  for (let i = 0; i < size; i++) {
    rows.push({ name: 'hello_' + i });
  }
  return rows;
}
