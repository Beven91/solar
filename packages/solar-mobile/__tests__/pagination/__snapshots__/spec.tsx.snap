// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination pagination 1`] = `
<div
  className="reverser-scroll order-list pagination-view"
  onScroll={[Function]}
>
  <div
    className="reverse-header"
  >
    <div
      className="scroll-end-loading"
    >
      <svg
        className="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          xlinkHref="#loading"
        />
      </svg>
      <span
        className="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
  <div>
    <div
      className="row-2"
    >
      <div>
        hello_2
      </div>
    </div>
    <div
      className="row-1"
    >
      <div>
        hello_1
      </div>
    </div>
    <div
      className="row-0"
    >
      <div>
        hello_0
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination bottom-to-top 1`] = `
<div
  className="reverser-scroll reverse-pagination pagination-view"
  onScroll={[Function]}
>
  <div
    className="reverse-header"
  >
    <div
      className="scroll-end-loading"
    >
      <svg
        className="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          xlinkHref="#loading"
        />
      </svg>
      <span
        className="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
  <div>
    <div
      className="row-2"
    >
      <div>
        hello_2
      </div>
    </div>
    <div
      className="row-1"
    >
      <div>
        hello_1
      </div>
    </div>
    <div
      className="row-0"
    >
      <div>
        hello_0
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination custom scroller 1`] = `
<div
  className="order-list pagination-view"
>
  <div
    className="am-list-body"
  >
    <div
      className="list-view-section-body"
    >
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
    </div>
  </div>
  <div
    className="am-list-footer"
  >
    <div
      className="scroll-end-loading"
    >
      <svg
        className="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          xlinkHref="#loading"
        />
      </svg>
      <span
        className="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination custom scroller 2`] = `
<div
  class="order-list pagination-view"
>
  <div
    class="am-list-body"
  >
    <div
      class="list-view-section-body"
    >
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
    </div>
  </div>
  <div
    class="am-list-footer"
  >
    <div
      class="scroll-end-loading"
    >
      <svg
        class="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          href="#loading"
        />
      </svg>
      <span
        class="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination custom scroller 3`] = `
<div
  class="order-list pagination-view"
>
  <div
    class="am-list-body"
  >
    <div
      class="list-view-section-body"
    >
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
    </div>
  </div>
  <div
    class="am-list-footer"
  >
    <div
      class="scroll-end-loading"
    >
      <svg
        class="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          href="#loading"
        />
      </svg>
      <span
        class="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination increment 1`] = `
<div
  className="order-list pagination-view"
>
  <div
    className="am-list-body"
  >
    <div
      className="list-view-section-body"
    >
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
    </div>
  </div>
  <div
    className="am-list-footer"
  >
    <div
      className="scroll-end-loading"
    >
      <svg
        className="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          xlinkHref="#loading"
        />
      </svg>
      <span
        className="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination increment 2`] = `
<div
  className="order-list pagination-view"
>
  <div
    className="am-list-body"
  >
    <div
      className="list-view-section-body"
    >
      <div>
        hello_0
      </div>
      <div>
        hello_1
      </div>
      <div>
        hello_2
      </div>
    </div>
  </div>
  <div
    className="am-list-footer"
  >
    <div
      className="scroll-end-loading"
    >
      <svg
        className="loading-icon am-icon am-icon-loading am-icon-xxs"
      >
        <use
          xlinkHref="#loading"
        />
      </svg>
      <span
        className="loading-text"
      >
        加载中，请稍后....
      </span>
    </div>
  </div>
</div>
`;

exports[`Pagination pagination.componentWillReceiveProps 1`] = `
<div
  class="pagination-view undefined list-empty"
/>
`;

exports[`Pagination pagination-revese 1`] = `
<div
  className="order-list pagination-view am-list-view-scrollview"
  style={
    Object {
      "WebkitOverflowScrolling": "touch",
      "overflow": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="am-list-view-scrollview-content am-list"
    style={
      Object {
        "minWidth": "100%",
        "position": "absolute",
      }
    }
  >
    <div
      className="am-list-body"
    >
      <div
        className="list-view-section-body"
      >
        <div>
          hello_0
        </div>
        <div>
          hello_1
        </div>
        <div>
          hello_2
        </div>
      </div>
    </div>
    <div
      className="am-list-footer"
    >
      <div
        className="scroll-end-loading"
      >
        <svg
          className="loading-icon am-icon am-icon-loading am-icon-xxs"
        >
          <use
            xlinkHref="#loading"
          />
        </svg>
        <span
          className="loading-text"
        >
          加载中，请稍后....
        </span>
      </div>
    </div>
  </div>
</div>
`;
